File: C#/Solutions/Problem020/SinglyLinkedList.cs
namespace DailyCodingProblems.Solutions.Problem020
{
    /// <summary>
    /// Singly linked list
    /// </summary>
    /// <typeparam name="T"></typeparam>
    public class SinglyLinkedList<T>
    {
        private SinglyLinkedListNode _head;
        public int Length { get; private set; }

        /// <summary>
        /// Adds a <see cref="SinglyLinkedListNode"/> as last element in the <see cref="SinglyLinkedList{T}"/>
        /// </summary>
        /// <param name="value"></param>
        public void AddLast(T value)
        {
            if (_head == null)
            {
                _head = new SinglyLinkedListNode(value);
            }
            else
            {
                var current = _head;
                while (current.NextNode != null)
                {
                    current = current.NextNode;
                }
                current.NextNode = new SinglyLinkedListNode(value);
            }
            Length++;
        }

        /// <summary>
        /// Gets the first <see cref="SinglyLinkedListNode"/> from a <see cref="SinglyLinkedList{T}"/>
        /// </summary>
        /// <returns></returns>
        public SinglyLinkedListNode GetFirst()
        {
            return _head;
        }

        /// <summary>
        /// Node in a <see cref="SinglyLinkedList{T}"/>
        /// </summary>
        public class SinglyLinkedListNode
        {
            public T Value { get; }
            public SinglyLinkedListNode NextNode { get; set; }

            public SinglyLinkedListNode(T value)
            {
                Value = value;
                NextNode = null;
            }
        }
    }
}

File: Java/Solutions/Problem020/SinglyLinkedList.java
package Problem020;

/**
 * Singly linked list
 */
public class SinglyLinkedList<T> {
    private SinglyListNode _head;
    private int length;

    public int getLength() {
        return length;
    }

    /**
     * Adds a node at the end of the singly linked list
     * @param value
     */
    public void addLast(T value) {
        if (_head == null){
            _head = new SinglyListNode(value);
        }
        else {
            var current = _head;
            while (current.nextNode != null) {
                current = current.nextNode;
            }
            current.nextNode = new SinglyListNode(value);
        }
        length++;
    }

    /**
     * Get the first element of the singly linked list
     */
    public SinglyListNode getFirst() {
        return _head;
    }

    /**
     * Node in a singly linked list
     */
    public class SinglyListNode {
        private T value;
        private SinglyListNode nextNode;

        public SinglyListNode(T value) {
            this.value = value;
        }

        public T getValue() {
            return value;
        }

        public SinglyListNode getNextNode() {
            return nextNode;
        }

        public void setNextNode(SinglyListNode nextNode) {
            this.nextNode = nextNode;
        }
    }
}
